relayer/crates/starknet-chain-components/src/impls/messages/ibc_transfer.rs

```diff
@@ -58,91 +56,126 @@ where
-            .map_err(Chain::raise_error)?;
-
-        let output = chain
-            .call_contract(
-                &ics20_contract_address,
-                &selector!("ibc_token_denom"),
-                &calldata,
-                None,
-            )
-            .await?;
-
-        let prefix_denom_str = chain
-            .encoding()
-            .decode(&output)
-            .map_err(Chain::raise_error)?;
-
-        let denom = PrefixedDenom::from_str(&prefix_denom_str).map_err(Chain::raise_error)?;
-
-        let timeout_height_on_b = if let Some(timeout_height) = timeout_height {
-            Height {
-                revision_number: 0,
-                revision_height: *timeout_height,
-            }
-        } else {
-            Height {
-                revision_number: 0,
-                revision_height: 0,
-            }
-        };
+        let approve_message = {
+            let call_data = encoding
+                .encode(&product![ics20_contract_address, amount.quantity,])
+                .map_err(Chain::raise_error)?;
 
-        let timeout_timestamp_on_b = if let Some(timeout_time) = timeout_time {
-            *timeout_time
-        } else {
-            Timestamp::from_nanoseconds(0)
-        };
+            let call = Call {
+                to: amount.token_address.0,
+                selector: selector!("approve"),
+                calldata: call_data,
+            };
 
-        let memo = if let Some(memo) = memo {
-            memo.clone()
-        } else {
-            String::new()
+            StarknetMessage::new(call)
         };
 
-        let ics20_transfer_message = MsgTransfer {
-            port_id_on_a: port_id.clone(),
-            chan_id_on_a: channel_id.clone(),
-            denom,
-            amount: amount.quantity,
-            receiver: recipient_address.to_string(),
-            memo,
-            timeout_height_on_b,
-            timeout_timestamp_on_b,
-        };
+        let transfer_message = {
+            let calldata = chain
+                .encoding()
+                .encode(&product![amount.token_address])
+                .map_err(Chain::raise_error)?;
+
+            let denom = {
+                let output = chain
+                    .call_contract(
+                        &ics20_contract_address,
+                        &selector!("ibc_token_denom"),
+                        &calldata,
+                        None,
+                    )
+                    .await;
+
+                match output {
+                    Ok(output) => {
+                        let prefix_denom_str =
+                            encoding.decode(&output).map_err(Chain::raise_error)?;
+
+                        PrefixedDenom::from_str(&prefix_denom_str).map_err(Chain::raise_error)?
+                    }
+                    Err(_) => {
+                        // FIXME: check and verify error is "ICS20: missing token denom", which means
```

Comment:
The error could be logged as warning in the meantime

Reply:
I have modified the Cairo contract to return `Option` in 63d947c5fdc822b1d66256d9118fb7768372c8f6, so that we can pattern match on the response in case the denom is not available.

Reply:
This is now done in d91106734d01255c888a589f4baba443d7868afc.